crates/cxx-qt-lib/src/types/qvariant/mod.rs

```diff
@@ -0,0 +1,167 @@
+
+        /// Convert this variant to type QMetaType::UnknownType and free up any resources used.
+        fn clear(self: &mut QVariant);
+        /// Returns true if this is a null variant, false otherwise.
+        #[rust_name = "is_null"]
+        fn isNull(self: &QVariant) -> bool;
+        /// Returns true if the storage type of this variant is not QMetaType::UnknownType; otherwise returns false.
+        #[rust_name = "is_valid"]
+        fn isValid(self: &QVariant) -> bool;
+    }
+
+    #[namespace = "rust::cxxqtlib1"]
+    unsafe extern "C++" {
+        include!("cxx-qt-lib/common.h");
+
+        #[doc(hidden)]
+        #[rust_name = "qvariant_drop"]
+        fn drop(variant: &mut QVariant);
+        #[doc(hidden)]
+        #[rust_name = "qvariant_default"]
+        fn construct() -> QVariant;
+        #[doc(hidden)]
+        #[rust_name = "qvariant_clone"]
+        fn construct(variant: &QVariant) -> QVariant;
+    }
+}
+
+/// The QVariant class acts like a union for the most common Qt data types.
+#[repr(C)]
+pub struct QVariant {
+    /// The layout has changed between Qt 5 and Qt 6
+    ///
+    /// Qt5 QString has one member, which contains three uints (but they are optimised to a size of 8) and a union
+    /// Qt6 QString has one member, which contains three pointers and a union (pointer largest)
+    #[cfg(qt_version_major = "5")]
+    _space: MaybeUninit<[usize; 2]>,
+    #[cfg(qt_version_major = "6")]
+    _space: MaybeUninit<[usize; 4]>,
+}
+
+impl Clone for QVariant {
+    /// Constructs a copy of the variant, p, passed as the argument to this constructor.
+    fn clone(&self) -> Self {
+        ffi::qvariant_clone(self)
+    }
+}
+
+impl Default for QVariant {
+    /// Constructs an invalid variant.
+    fn default() -> Self {
+        ffi::qvariant_default()
+    }
+}
+
+impl Drop for QVariant {
+    /// Destroys the QVariant and the contained object.
+    fn drop(&mut self) {
+        ffi::qvariant_drop(self)
+    }
+}
+
+impl<T> From<&T> for QVariant
+where
+    T: QVariantValue,
+{
+    /// Constructs a QVariant from a value of T
+    fn from(value: &T) -> Self {
+        T::construct(value)
+    }
+}
+
+// Note we can't use impl Into or TryInto for QVariant here as it conflicts
+//
+// note: conflicting implementation in crate `core`:
+// - impl<T, U> TryInto<U> for T
+//   where U: TryFrom<T>;
+impl QVariant {
+    /// Returns the stored value converted to the template type T
+    /// if QVariant::canConvert is true otherwise returns None
+    pub fn try_value<T: QVariantValue>(&self) -> Option<T> {
+        if T::can_convert(self) {
+            Some(T::value(self))
+        } else {
+            None
+        }
+    }
+
+    /// Returns the stored value converted to the template type T
+    ///
+    /// If the value cannot be converted, a default-constructed value will be returned.
+    pub fn value<T: QVariantValue>(&self) -> T {
+        T::value(self)
+    }
+}
+
+pub trait QVariantValue: Sized {
```

Comment:
should add more docs around how to use this. Likely also need to do the same for QSet/QHash etc?

Reply:
Added an example of a implementing a custom struct with QVariant too.